### PR TITLE
feat: add pipeline analytics

### PR DESCRIPTION
**Pull request by chaynabors**

### CODE REVIEW

File: crates/cli/src/main.rs [hunk 1]
```diff
@@ -495,7 +513,7 @@ impl Commands {
                     .collect();
                 let graph = topological_sort(&invert_graph(&nodes))?;
 
-                for run_group in graph {
+                'run_groups: for run_group in graph {
```
Comment (grant0417) [inline crates/cli/src/main.rs:516]: Do we ever exit `exit 1` any more?

File: crates/cli/src/main.rs [hunk 2]
```diff
@@ -375,6 +375,24 @@ impl Commands {
 
                 let pipeline = resolve_pipeline(pipeline)?;
                 let pipeline_file_name = pipeline.file_name().unwrap();
+
+                #[cfg(feature = "telemetry")]
+                let telem_join = segment_enabled().then(|| {
+                    let pipeline_name = pipeline_file_name.to_string_lossy().to_string();
+                    let pipeline_length = std::fs::File::open(&pipeline)
+                        .map(|f| f.metadata().map(|m| m.len()).ok())
```
Comment (grant0417) [inline crates/cli/src/main.rs]: This bytes length, is that what we want?

**Review by grant0417 — CHANGES_REQUESTED** [review]

File: crates/cli/src/main.rs [hunk 3]
```diff
@@ -651,22 +669,31 @@ impl Commands {
...
                         Err(e) => bail!(e),
                     }
                 }
+
+                #[cfg(feature = "telemetry")]
+                if let Some(join) = telem_join {
+                    join.await.ok();
+                }
+
+                std::process::exit(exit_code)
```
Comment (grant0417) [inline crates/cli/src/main.rs]: Can you do this only if `!= 1` slightly pedantic :)

Comment (chaynabors) [reply crates/cli/src/main.rs]: thought about it, sure

File: crates/cli/src/main.rs [hunk 4]
```diff
@@ -375,6 +375,23 @@ impl Commands {
 
                 let pipeline = resolve_pipeline(pipeline)?;
                 let pipeline_file_name = pipeline.file_name().unwrap();
+
+                #[cfg(feature = "telemetry")]
+                let telem_join = segment_enabled().then(|| {
+                    let pipeline_name = pipeline_file_name.to_string_lossy().to_string();
```
Comment (grant0417) [inline crates/cli/src/main.rs:381]: Lets make this a salted hash!
- 1. store a local salt like anon id
- 2. make a sha of the pipeline name with the salted hash simple as `hash(format!("{name}{salt}"))`
- 3. profit by not leaking names!

Comment (grant0417) [reply crates/cli/src/main.rs:381]: Also use sha256

File: crates/cli/Cargo.toml [hunk 5]
```diff
@@ -58,6 +58,7 @@ sentry = { version = "0.30.0", default-features = false, features = [
 ], optional = true }
 serde = { version = "1.0.159", features = ["derive"] }
 serde_json = "1.0.94"
+sha256 = "1.1.2"
```
Comment (grant0417) [inline crates/cli/Cargo.toml]: https://crates.io/crates/sha2

File: crates/cli/src/telemetry/segment.rs [hunk 6]
```diff
@@ -49,6 +71,55 @@ impl TrackEvent {
...
                     .collect::<Map<String, Value>>(),
             ),
+            TrackEvent::PipelineExecuted {
+                pipeline_name,
+                pipeline_length,
+            } => (
+                "pipeline_executed".into(),
+                [
+                    (
+                        "pipeline_name".to_owned(),
```
Comment (grant0417) [inline crates/cli/src/telemetry/segment.rs]:
```suggestion
                        "pipeline_name_hash".to_owned(),
```

**Review by grant0417 — CHANGES_REQUESTED** [review]
a

**Pull request merged**